docs/composition.md

```diff
@@ -0,0 +1,882 @@
+---
+title: Publishing Infrastructure
+toc: true
+weight: 400
+indent: true
+---
+
+# Publishing Infrastructure
+
+Crossplane allows infrastructure operators to define new kinds of infrastructure
+resources and publish them for the application operators they support to use,
+all without writing any code.
+
+Infrastructure providers enable Crossplane to manage a wide array of
+infrastructure resources like Azure SQL servers and AWS ElastiCache clusters.
+Infrastructure composition allows infrastructure operators to define, share, and
+reuse new kinds of infrastructure resources that are _composed_ of other
+infrastructure resources. Infrastructure operators may configure one or more
+compositions of any defined resource, and may publish any defined resource to
+their application operators, who may then declare that their application
+requires that kind of resource.
+
+Composition can be used to build a catalogue of kinds and configuration classes
+of infrastructure that fit the needs and opinions of your organisation. As an
+infrastructure operator you might define your own `MySQLInstance` resource. This
+resource would allow your application operators to configure only the settings
+that _your_ organisation needs - perhaps engine version and storage size. All
+other settings are deferred to a selectable composition representing a
+configuration class like "production" or "staging".
+
+> Note that composition is an **experimental** feature of Crossplane. Refer to
+> [Current Limitations] for information on functionality that is planned but not
+> yet implemented.
+
+## Concepts
+
+![Infrastructure Composition Concepts]
+
+A _Composite_ infrastructure resource is composed of other resources. Its
+configuration schema is user-defined. The `MySQLInstance` resources in the above
+diagram are composite infrastructure resources.
+
+A `Composition` specifies a set of resources of which a composite infrastructure
+resource may be composed. The Azure and AWS `Composition` resources in the above
+diagram configure two possible compositions of a `MySQLInstance`.
+
+A _Requirement_ for an infrastructure resource declares that an application
+requires a particular kind and configuration of composite infrastructure
+resource. The `MySQLInstanceRequirement` resources in the above diagram declare
+that the application pods each require a `MySQLInstance`.
+
+An `InfrastructureDefinition` defines a new kind of composite infrastructure
+resource. The `InfrastructureDefinition` in the above diagram defines the
+existence and schema of the `MySQLInstance` composite infrastructure resource.
+
+An `InfrastructurePublication` publishes a defined composite infrastructure
+resource, allowing application operators to declare a requirement for it. The
+`InfrastructurePublication` in the above diagram allows application operators to
+author a `MySQLInstanceRequirement`.
```

Comment:
This paragraph leaves me wondering why Publications exist? Maybe we could explain why creating an `InfrastructureDefinition` doesn't inherently publish it, or why there isn't just a `published` flag on the definition.

Reply:
We've had similar thoughts, especially because using a separate publication resource adds some hairy implementation details - e.g. you need to delete the publication before deleting the definition.

We've already found some things that we want to configure at the InfrastructureDefinition level that aren't just a CRD template - e.g. the connection secret details a particular kind of infrastructure needs. No such things exist for publications today, but if we think they might in future it's probably worth keeping it as a distinct resource. If we don't, I'd err toward making publication a flag on the InfrastructureDefinition. @muvaf has pointed out that we could let RBAC decide whether folks should be able to author a requirement, but I like the UX of being able to `kubectl get infrastructurepublications` to discover which resources are published and which are not. `kubectl get infrastructuredefinitions` with a `PUBLISHED` column would likely be just as good.

Either way, I'm not sure user facing documentation is the right place for us to be justifying the value/existence of a concept. It exists, and it will exist for at least Crossplane 0.11, so I think we should document what it does today.